bumble/smp.py

```diff
@@ -1098,15 +1098,47 @@ def start_encryption(self, key: bytes) -> None:
             )
         )
 
-    async def derive_ltk(self) -> None:
-        link_key = await self.manager.device.get_link_key(self.connection.peer_address)
-        assert link_key is not None
+    @classmethod
+    def derive_ltk(cls, link_key: bytes, ct2: bool) -> bytes:
+        '''Derives Long Term Key from Link Key.
+
+        Args:
+            link_key: BR/EDR Link Key bytes in little-endian.
+            ct2: whether ct2 is supported on both devices.
+        Returns:
+            LE Long Tern Key bytes in little-endian.
+        '''
+        link_key_big_endian = bytes(reversed(link_key))
         ilk = (
-            crypto.h7(salt=SMP_CTKD_H7_BRLE_SALT, w=link_key)
-            if self.ct2
-            else crypto.h6(link_key, b'tmp2')
+            crypto.h7(salt=SMP_CTKD_H7_BRLE_SALT, w=link_key_big_endian)
```

Comment:
I would prefer to update `h6` and `h7` to be implemented the same way the other bt crypto functions are done in `crypto.py`, where the inputs and outputs are in little-endian, and the byte reversing is done internally, as appropriate, when calling internal crypto primitives that are not BT specific (like `aes_cmac`). See `crypto.l6` for example. 
Also, I realize that `crypto.py` uses the expression `bytes(reversed(xyz))` a lot. Since you're editing that part of the code, it might be a good opportunity to use `xxx[::-1]` as a faster way to reverse a `bytes` value, or maybe even make that a local function `def reverse(x: bytes) -> bytes: return x[::-1]` in `crypto.py` for clarity.

Reply:
One problem here is that the spec provides a sample data for the derive procedure and I would like to test it. However, it would be strange if we just reimplement the same code in test.

Reply:
The other tests in `smp_tests.py` also use the test vectors from the spec, exactly as they are given (i.e big-endian), and reverses the bytes before calling the crypto functions (for example, look at `test_f6`, which uses the test vectors from `D.4 f6 LE SC CHECK VALUE GENERATION FUNCTION` in the spec. By having the `crypto.xxx` functions take their inputs and outputs in little-endian, it makes the code in `smp.py` simpler because it doesn't have to do the byte reversing each time before and after calling the crypto functions (see `smp.py` line 1711 and 1712 for example where `crypto.f6` is called.

Reply:
Okay, so should we keep a standalone `derive_[ltk/lk]` method? Also shall we move them to `crypto`?